[Step1] 날씨의아이

July911 opened the pull request:
안녕하세요 숲재 

외부라이브러리 없이 MVVM 클린아키텍쳐를 구현해보았습니다. 


## 파일 분리


![스크린샷 2022-04-13 오전 12 03 19](https://user-images.githubusercontent.com/90888402/162993129-986e84a6-c1dc-4c68-b98c-86f37ee3fc88.jpg)


## API 통신 `Generic` 구현 

```swift 
final class APIService<T: Decodable> {
        
    func request(
        _ type: RequestType,
        completion: @escaping (Result<T, Error>) -> Void
    ) {
```

각각의 레포지토리에서 APIService를 제네릭으로 가지고 있어서 나중에 다른 URL로 통신을 하게 될 경우에도 재사용 할 수 있게 했습니다.


## 레포지토리에서 데이터 변환 

```swift 
func getWeatherInformation(cityName text: String, completion: @escaping (TodayWeather?) -> Void) {
        let request: RequestType = .getWeatherFromCityName(city: text)
        
        service.request(request) { result in
            switch result {
            case .success(let weatherInformation):
                let todayWeather = weatherInformation.toDomain()
                completion(todayWeather)
            case .failure(_):
                completion(nil)
            }
        }
    }
``` 

API통신을 통해 온 `Decodable` 을 채택한 `Entity`인 `WeatherInformation`에서 앱 전반적으로 쓰이는 `TodayWeather`로 바꿔주었습니다.
Service (API통신) 단에서 변환을 해주어야할지 레포지토리에서 바꿔주어서 `UseCase`에 보내야할지 헷갈립니다 ! 숲재의 의견은 어떠신가요 ?
 
### 수동적인 뷰컨 구성 

![스크린샷 2022-04-13 오전 12 14 37](https://user-images.githubusercontent.com/90888402/162995421-4f35667d-0809-4d17-8c6b-2b9d9c7f63fb.jpg)

`ViewController`를 최대한 수동적으로 짜보려고 했습니다. 
하지만, searchBar가 active 상태인지, 비었는지를 확인해주는 `isFiltering` 연산 프로퍼티를 만들어서 사용하고 있는데 , 
이걸 비지니스 로직이라고 봐야할까요 아니면 여기다 그냥 둬도 되는걸까요 ? 
숲재의 생각이 궁금합니다 !

July911 committed:
Initial Commit

July911 committed:
Add files via upload

initial commit

July911 committed:
feat: Storyboard 삭제 및 파일정리

July911 committed:
feat: NetWork 세부사항 구현

- API Request 별로 fullURL을 구성하는 enum 구현
- 불변의 BaseURL 과 APIKey를 담는 Constant 열거형 구현

July911 committed:
feat: URLSession을 사용한 네트워킹 구현

July911 committed:
feat: Weather Information 타입 생성

July911 committed:
refactor: 필요없는 주석 제거 및 Final 키워드 추가

July911 committed:
feat: 날씨정보를 불러올 수 있는 Repository 생성

July911 committed:
feat: Asset에서 City의 정보를 불러올 수 있는 Repository생성

July911 committed:
feat: City를 찾을 수 있는 UseCase 생성

July911 committed:
feat: DetailShowUseCase 생성

July911 committed:
chore: 불필요한 주석 제거

July911 committed:
feat: UIComponents 추가

- SearchViewController 에 TableView 추가
- ListTableViewCell 생성
- TableView Delegate, DataSource를 ViewController 에 채택
- ViewModel을 SearchViewController 내부에 생성

July911 committed:
feat: Coordinator 구현 및 DetailShowViewController 구현

July911 committed:
feat: SearchController 구현

July911 committed:
refactor: Delegate 패턴 방향 수정

- VC의 Delegate를 VM 의 Delegate로 수정

July911 committed:
feat: ViewModelDelegate 구현

July911 committed:
feat: SearchController 의 searchResultsUpdater를 VC로 지정

July911 committed:
feat: Repository에서 Data원본을 받아오는 메서드 구현

July911 committed:
refactor: DetailShowUIViewController에 final 키워드 추가

July911 committed:
feat: GoogleMap WebView를 위한 URL추가

July911 committed:
feat: DetailViewController에 WebView추가

July911 committed:
feat: WebView Load를 위한 메서드 및 타입 추가

- 위경도로 주소를 알려주는 LocationSearchUseCase 생성
- WeatherRepository 내부에 Google에 주소를 검색한 URL를 생성하는 메서드 추가
- DetailShowViewModel 이 City의 정보를 받도록 로직 수정

July911 committed:
feat: DetailShowViewController Delegate 에 loadWebView 메서드 추가

- LocationSearchUseCase내부 메서드 search의 로직 탈출 클로저로 수정
- DetailShowViewModel 내부 createURL 메서드 구현
- Delegate를 통해 WebView가 Load되게 구현

July911 committed:
refactor: URL이 제대로 변환되지 않던 버그 수정

- " " 을 %20으로 수정
- BaseURL의 주소를 search에서 maps로 변경
- Locale을 ko-KR 에서 en-US로 변경

July911 committed:
feat: ListTableView의 layout 수정 및 위경도를 표시하게 구현

July911 committed:
refactor: 강제 언레핑 수정

July911 committed:
feat: DetailShowView UI component 추가

- Present 스타일을 FullScreen으로 변경
- 스택뷰로 전체 뷰를 묶고 아래 날씨정보를 표시하는 TextView추가

July911 committed:
feat: 데이터를 TextView에 전달

July911 committed:
feat: DetailShowViewController에 NavigationItem 추가

July911 committed:
feat: DefaultImageProvideRepository 구현

- NSCache를 사용한 캐싱 구현
- ImageCacheUseCase 생성 및 DetailShowViewModel 참조
- WebView SnapShot 구현

July911 committed:
feat: Image Cache 구현

- Asset Data에 불필요한 데이터 삭제
- DetailShowViewController 내부 Navigation RightBarButton 추가
- 스크린샷 버튼을 눌렀을때 캐싱되게 구현

July911 committed:
chore: SearchViewModel 프로퍼티 순서 변경

July911 committed:
chore: 필요없는 주석 제거

July911 committed:
feat: Weather Entity 생성

July911 committed:
feat: TableView의 estimatedRowHeight 지정

July911 committed:
feat: SearchViewController backgroundColor white로 변경

July911 committed:
faet: CleanArchitecture에 따라 model과 entity 분리

- DomainLayer에 model을 TodayWeather로 변경
- entity의 extension에toDomain메서드 추가
- view에 띄어줄 정보를 viewModel에서 변환

July911 committed:
refactor: ImageCacheUseCase에서 UIKit import 제거

- ImageCacheData 객체를 넘겨주는 방법으로 UIKit 제거
- ImageCacheUseCase를 ViewModel 에서 private 접근제어자 설정

July911 committed:
chore: 컨벤션 수정

July911 committed:
chore: 띄어쓰기 컨벤션 수정

July911 committed:
refactor: 중복된 프로젝트 파일 제거

forestjae commented on ChildOfWeather/Domain/RepositoryInterface/WeatherRepository.swift:
```diff
@@ -0,0 +1,11 @@
+        completion: @escaping (TodayWeather?) -> Void
+    )
+    
+    func getURLFromLoaction(locationAdress adress: String) -> String
```
오타발견

July911 replied:
수정했습니다 .. 민망하넹요

forestjae commented on ChildOfWeather/Domain/RepositoryInterface/WeatherRepository.swift:
```diff
@@ -0,0 +1,11 @@
+
+protocol WeatherRepository {
+    
+    func getWeatherInformation(
```
get네이밍은 지양하는걸로 배웠던것 같아요

July911 replied:
가져오다라는 동의어인 fetch 를 사용하여 바꿔주었습니다 !

forestjae commented on ChildOfWeather/Domain/UseCase/DetailShowUseCase.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+    
+    func getURL(from cityAdress: String) -> String {
+        self.weatherRepository.getURLFromLoaction(locationAdress: cityAdress)
```
get!

forestjae commented on ChildOfWeather/Domain/UseCase/DetailShowUseCase.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+}
+
+enum dataError: Error {
```
타입명!

July911 replied:
```swift
enum DataError: Error {
    case dataNotCome
}
``` 
수정했습니다 !

forestjae commented on ChildOfWeather/Domain/UseCase/LocationSearchUseCase.swift:
```diff
@@ -0,0 +1,22 @@
+        completion: @escaping (String?) -> Void
+    ) {
+        let findLocation = CLLocation(latitude: latitude, longitude: longitude)
+        let geocoder = CLGeocoder()
+        let locale = Locale(identifier: "en-US")
```
사용할때마다 매번 초기화가 되면 성능상 불리할것같아요

July911 replied:
repository 단의 initializer에서 CLGeocode 를 초기화해주고 해당 코드를 Coordinator의 프로퍼티로 옮겨 Coordinator가 SceneDelegate에서 초기화되는 시점에 한번만 생성되게 바꿔주었습니다. 
위경도에 따라 영어로 된 주소를 받아오는 작업은 ListTableView가 클릭이 되는 시점마다 발생하는 이벤트라 이렇게 처리해주었습니다.

forestjae commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -0,0 +1,53 @@
+        let request = URLRequest(url: url)
+
+        let task = URLSession.shared.dataTask(with: request) { data, URLResponse, _ in
+            
```
error 처리가 필요해보입니당

July911 replied:
```swift
            guard error == nil
            else {
                completion(.failure(APICallError.errorExist))
                return
            }
            
            guard let data = data
            else {
                completion(.failure(APICallError.dataNotfetched))
                return
            }
            
            guard let response = URLResponse as? HTTPURLResponse
            else {
                completion(.failure(APICallError.invalidResponse))
                return
            }
            
            guard (200...299) ~= response.statusCode
            else {
                completion(.failure(APICallError.notProperStatusCode))
                return
            }
            
            guard let decodedObject = try? JSONDecoder().decode(T.self, from: data)
            else {
                completion(.failure(APICallError.failureDecoding))
                return
            }
``` 
각 상황별로 에러처리를 해주었습니다

forestjae commented on ChildOfWeather/NetWork/Protocol/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,11 @@
+
+protocol URLSessionNetworkService {
+    
+    associatedtype decodedType: Decodable
```
컨벤션!

forestjae commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -0,0 +1,53 @@
+    case notProperStatusCode
+}
+
+final class APIService<T: Decodable> {
```
타입에 제네릭을 부여하는 것보다 메서드에 부여하는게 좀더 유연하게 서비스 객체를 쓸 수 있을것같아요

July911 replied:
```swift
  func request<T: Decodable>(
        decodedType: T.Type,
        requestType: RequestType,
        completion: @escaping (Result<T, APICallError>) -> Void
    ) 
``` 
타입에 부여하는 것이 아니라 메서드에 제네릭을 부여하여서 하나의 인스턴스로 다른 디코딩타입을 받아올 수 있게 구현하였습니다 !

forestjae commented on ChildOfWeather/NetWork/Protocol/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,11 @@
+    
+    func request(
+        _ type: RequestType,
+        completion: @escaping (Result<decodedType, Error>) -> Void
```
APIError Enum을 만들어두셨으니 타입도 명시해주는게?

July911 replied:
```swift
func request<T: Decodable>(
        decodedType: T.Type,
        requestType: RequestType,
        completion: @escaping (Result<T, APICallError>) -> Void
    )
``` 
명시해주었습니다 !

forestjae commented on ChildOfWeather/Presentation/DetailView/Utils/Double + Extension.swift:
```diff
@@ -0,0 +1,11 @@
+    
+    func toKoreanTime() -> String {
+        let time = Date(timeIntervalSince1970: TimeInterval(self))
+        let formatter = DateFormatter()
```
데이트 포매터 생성 비용이 크다고하네요

July911 replied:
```swift
class DefaultDateformatter {
    
    static let shared = DefaultDateformatter()
    let dateformatter = DateFormatter()
    
    private init() {
        
    }
}

``` 
공용 인스턴스로 만들어주었습니다 !

forestjae commented on ChildOfWeather/Presentation/DetailView/View/DetailShowUIViewController.swift:
```diff
@@ -0,0 +1,163 @@
+    
+    var viewModel: DetailShowViewModel?
+
+    private let WebView: WKWebView = {
```
컨벤션

forestjae commented on ChildOfWeather/Presentation/DetailView/View/DetailShowUIViewController.swift:
```diff
@@ -0,0 +1,163 @@
+
+final class DetailShowUIViewController: UIViewController {
+    
+    var viewModel: DetailShowViewModel?
+
```
뷰모델 추상화 추천드립니다.

July911 replied:
ViewModelProcol로 추상화 했습니다 !

forestjae commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+    private let detailShowUseCase: DetailShowUseCase
+    private let locationSearchUseCase: LocationSearchUseCase
+    private let imageCacheUseCase: ImageCacheUseCase
+    weak var delegate: DetailViewModelDelegate?
```
여기에 delegate가 있어야하는 이유가있을까요

forestjae commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+        self.imageCacheUseCase.getImage(cityName: key)
+    }
+    
+    func extractWeatherDescription() {
```
View입장에서 viewModel을 Delegate로 알고 있고, 이런 메서드들의 반환값을 받아서 view에서 변경을 처리하는게 맞지않을까 싶습니다.

forestjae commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+
+private extension Double {
+    
+    func toCelsius() -> Double {
+        return (self - 32)/1.8
+    }
+}
```
작업의 성격상 계산프로퍼티가 적절해 보입니다! 에러처리가 필요없고 사이드이펙트가 없으니 계산프로퍼티를 적극 활용하는 것이 좋아보여요.
https://ios-development.tistory.com/641
참고하세요

July911 replied:
계산 프로퍼티로 변경했습니다 :)

forestjae commented on ChildOfWeather/Presentation/SearchView/View/SearchViewController.swift:
```diff
@@ -0,0 +1,117 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView,didSelectRowAt indexPath: IndexPath) {
```
띄어쓰기가 빠졌어요

July911 replied:
수정했습니다 !

forestjae commented on ChildOfWeather/Presentation/SearchView/View/SearchViewController.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    func tableView(_ tableView: UITableView,didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: false)
+        let city = self.viewModel?.filterdResults?[indexPath.row] ?? City.EMPTY
```
nil이면 리턴 처리 하는게 좋지않을까요?

July911 replied:
맞네요 ... 리턴처리 해주었습니다 ! 감사합니다

forestjae commented:
수고하셨습니다. 내일  같이 이야기 해봐요.
깔끔하게 잘 짜신것 같아요.

## 통신 Generic 구현
코드에도 남겼는데, 타입자체가 제네릭으로 선언되면 이 타입으로는 한가지 엔티티밖에 못받아오는거잖아요? 향후 기능이 추가되서 리파지터리에서 다른 모델 타입을 파싱해오려면 별도의 APIService 인스턴스를 또 만들어야되는 문제가 발생할것 같아요.

## 레포지토리에서 데이터 변환
지금 하신 것처럼, 리파지터리에서 도메인에서 쓰는 모델 엔티티로 바꾸는 작업을 완료해야될 것 같습니다.

## 수동적인 뷰컨 구성
이정도는 괜찮을 것 같은데요? 다만 별도의 프로퍼티 없이 딜리게이트 함수 내부에서 아래 값만 체크하면 되는거 아닌가 하는 생각이 들어요
```swift
        let isSearchBarHasText = self.navigationItem.searchController?.searchBar.text?.isEmpty == false
```

July911 committed:
refactor: API서비스 Generic타입을 메서드로 Generic으로 변경

- 오탈자 수정
- 하나의 APIService 인스턴스에서 여러 디코딩타입을 호출할 수 있게 변경
- 컨벤션 수정

July911 committed:
refactor: Dateformatter 공용 인스턴스 생성

- fileprivate extension으로 구현된 데이트 포메팅 메서드를 연산프로퍼티로 변경
- Cell에서 위경도를 알려주는 configureCell에도 데이트 포메팅 적용
- Coordinator에서 함수 안에서 인스턴스 생성된 부분을 분리

July911 committed:
refactor: URLSession dataTask 내부에서 에러처리 세분화

July911 committed:
feat: Coordinator 접근제어자 private으로 설정

July911 committed:
refactor: ViewModel 을 protocol로 추상화
